x/btclightclient/keeper/keeper.go

```diff
@@ -99,13 +99,49 @@ func (k Keeper) insertHeaders(
 		k.triggerRollBack(ctx, result.RollbackInfo.HeaderToRollbackTo)
 	}
 
+	k.insertHeadersFromInsertResult(ctx, result, true)
+	return nil
+}
+
+func (k Keeper) insertHeadersNoEventsAndRollback(
```

Comment:
tbh there are 2 things about this approach which I do not like:
1. Part of insert result is handled in ` insertHeadersNoEventsAndRollback` and part of it is handled in `insertHeadersFromInsertResult`
2. `insertHeadersFromInsertResult` takes bool argument which is something to be avoided

What do you think about:
1. Introducing internal function:
```
func (k Keeper) insertHeadersInternal(
	ctx context.Context,
	headers []*wire.BlockHeader,
	handleInsertResult func(ctx context.Context, s headersState, result *types.InsertResult) error,
) error {

	headerState := k.headersState(ctx)

	result, err := k.bl.InsertHeaders(
		headerState,
		headers,
	)

	if err != nil {
		return err
	}

	return handleInsertResult(ctx, headerState, result)
}
```

then having two functions to create different insert result handlers:
- with hooks and events
- just with insertion and checking no rollback happened

The one with hooks and events
```
func (k Keeper) triggerEventAndHandleHooksHandler() func(ctx context.Context, s headersState, result *types.InsertResult) error {
	return func(ctx context.Context, s headersState, result *types.InsertResult) error {
		// if we have rollback, first delete all headers up to the rollback point
		if result.RollbackInfo != nil {
			// roll back to the height
			s.rollBackHeadersUpTo(result.RollbackInfo.HeaderToRollbackTo.Height)
			// trigger rollback event
			k.triggerRollBack(ctx, result.RollbackInfo.HeaderToRollbackTo)
		}

		for _, header := range result.HeadersToInsert {
			h := header
			s.insertHeader(h)
			k.triggerHeaderInserted(ctx, h)
			k.triggerRollForward(ctx, h)
		}
		return nil
	}
}

```

and the insertion one:
```
func (k Keeper) insertHandler() func(ctx context.Context, s headersState, result *types.InsertResult) error {
	return func(ctx context.Context, s headersState, result *types.InsertResult) error {
		// if we have rollback, first delete all headers up to the rollback point
		if result.RollbackInfo != nil {
			return fmt.Errorf("rollback should not happend %+v", result.RollbackInfo)
		}

		for _, header := range result.HeadersToInsert {
			h := header
			s.insertHeader(h)
		}
		return nil
	}
}
```

Then `insertWithHookAndEvents` becomes:
```
func (k Keeper) insertHeadersWithHookAndEvents(
	ctx context.Context,
	headers []*wire.BlockHeader) error {

	return k.insertHeadersInternal(
		ctx,
		headers,
		k.triggerEventAndHandleHooksHandler(),
	)
}
```

and option without events will be similiar but with different handler. 

This way both problems are fixed:
- handling the result is always in one function
- there is no bools

Reply:
understood, addressed at https://github.com/babylonlabs-io/babylon/pull/4/commits/c3de831e686256b621ad004f8f348c8b95c8b647 

Thanks for the suggestion